scripts/generate-plugins.go

```diff
@@ -0,0 +1,273 @@
+	}
+
+	log.Printf("Generated %s", target)
+}
+
+type plugin struct {
+	Package    string // This plugin's package name (aws)
+	PluginName string // Name of plugin (aws); important if there is a dash
+	TypeName   string // Type of plugin (provider)
+	Path       string // Path relative to terraform builtin (provider/aws)
+	ImportName string // PluginName+TypeName (awsprovider)
+}
+
+// makeProviderMap creates a map of providers like this:
+//
+// var InternalProviders = map[string]plugin.ProviderFunc{
+// 	"aws":        aws.Provider,
+// 	"azurerm":    azurerm.Provider,
+// 	"cloudflare": cloudflare.Provider,
+func makeProviderMap(varName, varType string, items []plugin) string {
+	output := ""
+	for _, item := range items {
+		output += fmt.Sprintf("\t\"%s\":   %s.%s,\n", item.PluginName, item.ImportName, item.TypeName)
+	}
+	return output
+}
+
+// makeProvisionerMap
+func makeProvisionerMap(varName, varType string, items []plugin) string {
+	output := ""
+	for _, item := range items {
+		// "local-exec":  func() terraform.ResourceProvisioner { return new(localexec.ResourceProvisioner) },
+		output += fmt.Sprintf("\t\"%s\":   func() terraform.ResourceProvisioner { return new(%s.%s) },\n", item.PluginName, item.ImportName, item.TypeName)
+	}
+	return output
+}
+
+func makeImports(providers, provisioners []plugin) string {
+	plugins := []string{}
+
+	for _, provider := range providers {
+		plugins = append(plugins, fmt.Sprintf("\t%s \"github.com/hashicorp/terraform/%s\"\n", provider.ImportName, filepath.ToSlash(provider.Path)))
+	}
+
+	for _, provisioner := range provisioners {
+		plugins = append(plugins, fmt.Sprintf("\t%s \"github.com/hashicorp/terraform/%s\"\n", provisioner.ImportName, filepath.ToSlash(provisioner.Path)))
+	}
+
+	// Make things pretty
+	sort.Strings(plugins)
+
+	return strings.Join(plugins, "")
+}
+
+// listDirectories recursively lists directories under the specified path
+func listDirectories(path string) ([]string, error) {
+	names := []string{}
+	items, err := ioutil.ReadDir(path)
+	if err != nil {
+		return names, err
+	}
+
+	for _, item := range items {
+		// We only want directories
+		if item.IsDir() {
+			if item.Name() == "test-fixtures" {
+				continue
+			}
+			currentDir := filepath.Join(path, item.Name())
+			names = append(names, currentDir)
+
+			// Do some recursion
+			subNames, err := listDirectories(currentDir)
+			if err == nil {
+				names = append(names, subNames...)
+			}
+		}
+	}
+
+	return names, nil
+}
+
+// deriveName determines the name of the plugin based on the path hierarchy:
+// provider/aws becomes awsprovider
+func deriveName(root, full string) string {
+	short, _ := filepath.Rel(root, full)
+	bits := strings.Split(short, string(os.PathSeparator))
+	return strings.Join(bits, "-")
+}
+
+// deriveImport will build a unique import identifier based on packageName and
+// the result of deriveName()
+//
+// This will be something like    -> awsprovider
```

Comment:
Yeah some of these are confusing thanks to some combination of copy+paste+find+replace. I will make a second pass here on comments.